### PR TITLE
fix: None() and bare none in branch-merge positions adopt contextual Option<T> inner type (#1154)

### LINKED ISSUE #1154: fix: None() outside return statement defaults to Option<i8> instead of contextual type
## Summary

`src/codegen_call.cpp` L783 付近で、`None()` を返り文以外の文脈（`if` 式、`case` 式、型注釈付き初期化子など）に使うと、囲む関数の返り値型から `T` が取れないため `Option<i8>` にハードコードされる。

## Repro (expected to fail)

```ry
fn f() -> Option<int>:
    x: Option<int> = None()   # このコンテキストで T が i8 になる
    x
```

## Expected

`None` リテラルと同じコンテキスト型付けパスを再利用し、型注釈・sibling ブランチから `T` を解決すべき。

## Fix direction

`emitNoneCall` に contextual type を渡すか、`None()` を `None` リテラルと統合する。

## Source

PR #1148 CodeRabbit review (src/codegen_call.cpp L783)

### PR DESCRIPTION
**Pull request by t0k0sh1**
## Summary

- Introduces a class-state **hint channel** (`option_none_hint_inner_` + `OptionNoneHintGuard` RAII guard) on `CodeGen` so `None()` and bare `none` in `if`/`case` arms adopt the sibling arm's `Option<T>` inner type instead of defaulting to `Option<i8>` or `Option<i64>`
- `computeBranchOptionInnerHint(a, b)` pre-scans both arms via `inferExprType` and picks the more-concrete inner type using `preferConcrete`
- Annotation seed at three `codegen_stmt.cpp` callsites ensures RHS `if`/`case` expressions with no sibling `Some(x)` arm still pick up the declared type
- ADR #1111 compliant — no `emitExpr` signature changes; hint is threaded via class state + RAII

## Affected files

| File | Change |
|---|---|
| `include/ry/codegen.hpp` | `option_none_hint_inner_` field, `OptionNoneHintGuard`, `computeBranchOptionInnerHint` decl |
| `src/codegen_lambda.cpp` | `computeBranchOptionInnerHint` impl; `NoneExpr` branch in `inferExprType` |
| `src/codegen_call.cpp` | `None()` emitter consults hint before `fn_->getReturnType()` |
| `src/codegen_expr.cpp` | `NoneExpr` emitter updated; `IfExpr`/`IfBlockExpr` pre-scan guard |
| `src/codegen_stmt.cpp` | Annotation seed at `emitVarDecl`, local reassign, module-global reassign |
| `tests/spec/option_branch_merge_none.test.ry` | 16 new spec cases |
| `changelog.d/1154-none-contextual-type.md` | Changelog fragment |
| `KNOWLEDGE.md` | Hint channel design and invariants documented |

## Known limitation

`None()` passed directly as a lambda call argument (e.g. `g(None())`) still defaults to `Option<i8>` because the argument-emit loop has no hint guard. Tracked separately in #1179.

## Test plan

- [x] 16 new `tests/spec/option_branch_merge_none.test.ry` cases cover `if`/`case`/nested/function-body positions for `T ∈ {int, str, List<int>}`
- [x] Existing `tests/spec/option_none_call.test.ry` (7 cases) continue to pass
- [x] Full test suite: `cmake --build build && ./build/ry_tests && ./build/ry test -p` — all pass
- [x] ASan + UBSan: clean
- [x] TSan: clean (C++ tests)
- [x] libFuzzer: `fuzz_parser` / `fuzz_json` / `fuzz_utf8` each 60 s — exit 0

Closes #1154

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixes type inference for `None()` and `none` in if/case branch-merge positions so they adopt the sibling arm’s concrete `Option<T>` inner type (including scrutinee/runtime-driven cases) instead of falling back to placeholder defaults.

* **Tests**
  * Added comprehensive regression tests covering `None`/`none` across if/case forms, nested branches, annotated all-None scenarios, multi-arm cases, and related edge cases.

* **Documentation**
  * Added changelog and knowledge base entries describing the fix, remaining limitation for function-argument positions, and rationale.

### CODE REVIEW

File: src/codegen_expr.cpp [hunk 1]
```diff
@@ -1746,6 +1746,12 @@ llvm::Value *CodeGen::emitExprVariant(const std::unique_ptr<CaseCondExpr> &e) {
 // ===== IfExpr (single-expression form: `if cond => then_val else else_val`) =====
 
 llvm::Value *CodeGen::emitExprVariant(const std::unique_ptr<IfExpr> &e) {
+    // Pre-scan: derive None() inner-type hint from the non-None sibling arm
+    // so both arms get the same Option<T> type (#1154). The guard restores
+    // the previous hint after this if-expr is fully emitted.
+    llvm::Type *ifExprHint = computeBranchOptionInnerHint(*e->then_value, *e->else_value);
+    OptionNoneHintGuard ifExprGuard(*this, ifExprHint ? ifExprHint : option_none_hint_inner_);
+
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_expr.cpp]: Fixed. `OptionNoneHintGuard` is now installed **after** the condition expression is fully emitted in both `IfExpr` and `IfBlockExpr`. The pre-scan (`computeBranchOptionInnerHint`) still runs before any emit to derive the hint, but the guard that writes `option_none_hint_inner_` is set up only once the condition IR has been generated. Same pattern applied to `CaseCondExpr`. Additionally, the hint field (`option_none_hint_inner_`) is separate from the declaration-annotation field (`option_decl_annotation_inner_`); the latter is written by `emitVarDecl`/reassign sites but is only used as a fallback inside the arm guards — it is never read directly by `None()`/`NoneExpr` emitters, so it cannot leak into function-argument sub-expressions.

File: src/codegen_lambda.cpp [hunk 2]
```diff
@@ -719,12 +719,53 @@ llvm::Type *CodeGen::inferExprType(const ExprNode &expr,
             return ptrTy_;
         } else if constexpr (std::is_same_v<T, std::unique_ptr<CastExpr>>) {
             return resolveType(v->target_type->toString());
+        } else if constexpr (std::is_same_v<T, NoneExpr>) {
+            // NoneExpr (bare `none`) is an Option literal with placeholder inner.
+            // Use i8Ty_ so preferConcrete() will prefer the sibling arm's inner type.
+            return getOptionType(i8Ty_);
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_lambda.cpp:725]: Fixed. Added a `NoneExpr` arm to `inferExprTypeName` (in `src/codegen_lambda.cpp`) that returns `"Option"`, consistent with `inferExprType` returning `getOptionType(i8Ty_)` for the same node. This ensures `computeBranchOptionInnerHint` and any callers that call `inferExprTypeName` on a `NoneExpr` see a coherent type string instead of falling through to the `reverseResolveTypeName` fallback.

File: src/codegen_lambda.cpp [hunk 3]
```diff
@@ -719,12 +719,53 @@ llvm::Type *CodeGen::inferExprType(const ExprNode &expr,
...
+        llvm::Type *hint = preferConcrete(inner(ta), inner(tb));
+        return (hint == i8Ty_) ? nullptr : hint;
+    }
+    if (aIsOpt) {
+        llvm::Type *ia = inner(ta);
+        return (ia == i8Ty_) ? nullptr : ia;
+    }
+    llvm::Type *ib = inner(tb);
+    return (ib == i8Ty_) ? nullptr : ib;
+}
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_lambda.cpp:767]: Addressed with a scrutinee-based hint approach in `src/codegen_match.cpp`. Pattern-match `case` arms reference pattern-bound variables (e.g., `v` in `Some(v)`) that are not in scope at pre-scan time, so `computeBranchOptionInnerHint` cannot be applied without producing incorrect types for non-integer inners. Instead, when the scrutinee type is `Option<T>`, its inner `T` is extracted and set as the `OptionNoneHintGuard` hint before any arm is emitted, with `option_decl_annotation_inner_` as the fallback. This covers `case o: Some(v) => Some(v * 2) None => None()` with an explicit annotation and lambda-return contexts. The KNOWLEDGE.md entry now documents this distinction.

File: src/codegen_stmt.cpp [hunk 4]
```diff
@@ -380,6 +380,18 @@ void CodeGen::emitVarDecl(const std::string &name,
...
+    // Option (not a literal/range constraint — those don't have inner types).
+    llvm::Type *annotOptionInner = nullptr;
+    if (annot && !constraint) {
+        const std::string &ra = resolvedAnnot.empty() ? *annot : resolvedAnnot;
+        llvm::Type *at = resolveType(ra);
+        if (isOptionType(at))
+            annotOptionInner = llvm::cast<llvm::StructType>(at)->getElementType(1);
+    }
+    OptionNoneHintGuard noneHintGuard(*this, annotOptionInner);
+
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_stmt.cpp:394]: Fixed via a two-field design. The declaration sites now write `option_decl_annotation_inner_` (via `DeclAnnotationInnerGuard`), not `option_none_hint_inner_`. The `None()`/`NoneExpr` emitters read only `option_none_hint_inner_` — they never consult `option_decl_annotation_inner_` directly. The declaration field is only picked up as a fallback inside `IfExpr`/`IfBlockExpr`/`CaseCondExpr` arm guards (i.e., after the condition has been emitted), so `none`/`None()` passed as a function argument within the same RHS tree never inherits the annotation hint. A regression test (`annotation seed does not leak into function arguments`) confirms `id_opt(none)` resolves independently.

File: tests/spec/option_branch_merge_none.test.ry [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  )
+)
+
+describe("nested if-expr with None() (#1154)", ():
+  it("should infer inner type through nested if", ():
+    outer: bool = true
+    inner_cond: bool = false
+    x: Option<int> = if outer => (if inner_cond => None() else Some(10)) else None()
+    expect(x).to_be_some()
+  )
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply tests/spec/option_branch_merge_none.test.ry:69]: Fixed. Added a `describe("None() in if-block-expr arms (#1154)")` group with two tests covering block-form `if` with `(None())` / `(Some(...))` arms. Note: in the `it()` lambda context, block-form `if` arms require parentheses around bare constructors (e.g., `(Some(1))`) because the `:` suffix is otherwise parsed as a block-end marker — the tests reflect this syntax requirement.

**Comment by t0k0sh1** [comment]
@coderabbitai

Re: KNOWLEDGE.md nitpick — "Add the known `None()` lambda-argument limitation to avoid overgeneralization."

Already addressed in the current KNOWLEDGE.md entry under **Known limitation**:

> `None()` and `none` in function argument positions (e.g., `f(none)`) do NOT inherit the enclosing declaration annotation; they fall back to `fn_->getReturnType()` or the default `i64`/`i8` placeholder. Filed as #1179. The `option_decl_annotation_inner_` field intentionally does NOT flow into argument positions to avoid incorrect type propagation.

This matches the suggested patch intent. No further change needed.

File: include/ry/codegen.hpp [hunk 6]
```diff
@@ -171,6 +171,53 @@ class CodeGen {
...
+    // emitters read it.
+    llvm::Type *option_none_hint_inner_ = nullptr;
+
+    // Declaration-context annotation inner type (#1154). Written by
+    // emitVarDecl / local-reassign / global-reassign to carry the LHS
+    // Option<T> inner into the RHS emitter. Read only by IfExpr/IfBlockExpr
+    // guards as a fallback when computeBranchOptionInnerHint returns nullptr.
+    // None()/NoneExpr emitters do NOT read this field directly, so it cannot
+    // leak into arbitrary sub-expressions (e.g., function arguments).
+    llvm::Type *option_decl_annotation_inner_ = nullptr;
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply include/ry/codegen.hpp:191]: Fixed. Updated the comment for `option_decl_annotation_inner_` in `include/ry/codegen.hpp` to list all guard paths that use this field as a fallback:

```
// Option<T> inner into the RHS emitter. Read by branch-arm guards
// (IfExpr, IfBlockExpr, CaseCondExpr, and CaseExpr scrutinee path) as a
// fallback when computeBranchOptionInnerHint or scrutinee-derived hints
// are unavailable.
```

Also applied a related structural improvement proactively: in `IfBlockExpr`, moved the `OptionNoneHintGuard` inside the `emitBodyTail` lambda (now takes a `tailHint` parameter), narrowing its scope to cover only the tail expression rather than the entire body including non-tail statements. This prevents any `None()`/`none` in non-tail variable declarations from inheriting the arm hint unexpectedly.

File: src/codegen_expr.cpp [hunk 7]
```diff
@@ -1703,6 +1703,14 @@ llvm::Value *CodeGen::emitBinaryOp(const std::string &op, llvm::Value *lhs, llvm
...
 
 llvm::Value *CodeGen::emitExprVariant(const std::unique_ptr<CaseCondExpr> &e) {
+    // Pre-scan: compute None() hint from first arm value and else_expr (#1154).
+    // Arm conditions are emitted before each value so we re-install the guard
+    // around each value emit to avoid leaking the hint into condition exprs.
+    llvm::Type *caseCondHint = nullptr;
+    if (!e->arms.empty())
+        caseCondHint = computeBranchOptionInnerHint(*e->arms[0].value, *e->else_expr);
+    llvm::Type *caseCondFallback = caseCondHint ? caseCondHint : option_decl_annotation_inner_;
+
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_expr.cpp]: Fixed. Changed the CaseCondExpr pre-scan in `src/codegen_expr.cpp` to iterate over all arm values instead of only `arms[0]`:

```cpp
llvm::Type *caseCondHint = nullptr;
for (const auto &arm : e->arms) {
    caseCondHint = computeBranchOptionInnerHint(*arm.value, *e->else_expr);
    if (caseCondHint) break;
}
```

Also added a regression test `"hint derived from non-first arm when arms[0] and else are None() (no annotation) (#1154)"` in `tests/spec/option_branch_merge_none.test.ry` that specifically exercises this path: `arms[0]` and `else_expr` are both `None()`, the concrete `Some(42)` is in `arms[1]`, and there is no LHS annotation to fall back on. With the single-arm scan this emitted `Option<i8>` vs `Option<int>` and failed branch validation; with the full scan all arms correctly emit `Option<int>`.

File: include/ry/codegen.hpp [hunk 8]
```diff
@@ -171,6 +171,55 @@ class CodeGen {
     // non-zero depth into subsequent unrelated codegen. See #630.
     int parallel_for_depth_ = 0;
 
+    // Hint for None()/none inner type in branch-merge positions (#1154).
+    // Set to the expected inner type before emitting an arm expression that
+    // may contain None() or bare none. nullptr means "no hint; fall back to
+    // fn_->getReturnType() or i8/i64 defaults".
+    // Only IfExpr/IfBlockExpr guards write this field; None()/NoneExpr
+    // emitters read it.
+    llvm::Type *option_none_hint_inner_ = nullptr;
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply include/ry/codegen.hpp:181]: Fixed. Updated the comment for `option_none_hint_inner_` in `include/ry/codegen.hpp` to list all writer paths:

```
// Written by branch-arm guard paths (IfExpr, IfBlockExpr, CaseCondExpr,
// and CaseExpr arm-scoped/scrutinee-derived flows); None()/NoneExpr
// emitters read it.
```

File: include/ry/codegen.hpp [hunk 9]
```diff
@@ -171,6 +171,55 @@ class CodeGen {
...
+            cg_.option_decl_annotation_inner_ = inner;
+        }
+        ~DeclAnnotationInnerGuard() { cg_.option_decl_annotation_inner_ = saved_; }
+        DeclAnnotationInnerGuard(const DeclAnnotationInnerGuard &) = delete;
+        DeclAnnotationInnerGuard &operator=(const DeclAnnotationInnerGuard &) = delete;
+    };
+
+    // Compute the Option inner type hint from a pair of if/case arm expressions.
+    // Returns nullptr if neither arm is an Option literal or both are placeholders.
+    llvm::Type *computeBranchOptionInnerHint(const ExprNode &a, const ExprNode &b);
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply include/ry/codegen.hpp]: Fixed. Updated the `computeBranchOptionInnerHint` docstring in `include/ry/codegen.hpp` to describe actual behavior:

```
// Compute an Option inner-type hint from a pair of branch-arm expressions.
// Returns nullptr when no concrete inner type can be derived
// (e.g., non-Option arms or placeholder-only outcomes).
```

"Option literal" was inaccurate — the function operates on any branch expression type, not only literal nodes.

**Comment by t0k0sh1** [comment]
@coderabbitai The KNOWLEDGE.md nitpick (lines 3315-3318: add `#1179` lambda-argument limitation) was addressed in commit `2d56eca2`. The "Known limitation" line was added directly below the CaseExpr entry:

> `None()` and `none` in function argument positions (e.g., `f(none)`) do NOT inherit the enclosing declaration annotation; they fall back to `fn_->getReturnType()` or the default `i64`/`i8` placeholder. Filed as #1179.
